slides/src/main/asciidoc/lecture.adoc

```diff
@@ -314,8 +326,48 @@ Kafka[shape=box,label="reduce \n (a, b) → <STOP>"]
 
 * In production it doesn't work
 
+
+== Why it's not working
+
+Kafka Streams is a big data streaming framework
+
+
+TopologyTestDriver is a fast, deterministic testing framework
+
+
+== Why it's not working
+
+Kafka Streams is a big data streaming framework
+* designed for high throughput
+* throughput demands batching, buffering, caching, etc.
+* caching is the culprit in this example
+
+
+TopologyTestDriver is a fast, deterministic testing framework
+
+
+== Why it's not working
+
+Kafka Streams is a big data streaming framework
+* designed for high throughput
+* throughput demands batching, buffering, caching, etc.
+* caching is the culprit in this example
+
+
+TopologyTestDriver is a fast, deterministic testing framework
+* designed for synchronous, immediate results
+* flush cache after every update
+
 == Why it's not working
 
+Caching in Kafka Streams
+* don't immediately emit every aggregation result
+* "soak up" repeated updates to the same key's aggregation
+* configure cache size: `max.bytes.buffering` (10MB)
+* configure cache flush interval: `commit.interval.ms` (30s)
+* emit _latest_ result on flush or eviction
```

Comment:
Added this sequence to give people the abstract concepts first so they can understand the example a little better.